File: examples/numpy_dag.cpp
//
// Created by astokely on 12/14/23.
//
#include "is_dag.h"
#include <filesystem>

int main() {
    int n, m;
    int **a;
    std::string path = "/home/astokely/CLionProjects/dag/numpy_test_graphs/bitarray_graphs";
    for (const auto & entry : std::filesystem::directory_iterator(path)) {
        std::cout
                << entry.path()
                << " ";
        loadNpy(
                entry.path(),
                &a,
                &m,
                &n
        );
        // print a

        std::cout
                <<
                isDAGBitArray(
                        (uint32_t
                        *) a[0],
                        m
                )
                <<
                std::endl;
        if (n) {
            for (
                    int i = 0;
                    i < m;
                    i++) {
                delete[] a[i];
            }
            delete[]
                    a;
        } else {
            delete[]
                    a[0];
            delete[]
                    a;
        }
    }
    path = "/home/astokely/CLionProjects/dag/numpy_test_graphs/matrix_graphs";
    for (const auto & entry : std::filesystem::directory_iterator(path)) {
        std::cout
                << entry.path()
                << " ";
        loadNpy(
                entry.path(),
                &a,
                &m,
                &n
        );

        std::cout
                <<
                isDAGAdjacencyMatrix(
                        (const int
                        **) a,
                        m

                )
                <<
                std::endl;
        if (n) {
            for (
                    int i = 0;
                    i < m;
                    i++) {
                delete[] a[i];
            }
            delete[]
                    a;
        } else {
            delete[]
                    a[0];
            delete[]
                    a;
        }
    }
    return 0;
}

File: examples/numpy_dag.py
from pathlib import Path

import numpy as np
import networkx as nx

from dag import isDAGAdjacencyMatrix as isDAGAdj
from dag import isDAGBitArray as isDAGBitArr
from dag import data_files

matrix_graphs = data_files("matrix_graphs")
bitarray_graphs = data_files("bitarray_graphs")
for graph in matrix_graphs:
    nx_graph = nx.from_numpy_array(np.load(graph), create_using=nx.DiGraph)
    print(graph, isDAGAdj(np.load(graph)), nx.is_directed_acyclic_graph(nx_graph))

for graph in bitarray_graphs:
    print(graph, isDAGBitArr(np.load(graph)))
